tests/Unit/Utilities/PregTest.php

```diff
@@ -0,0 +1,282 @@
+    {
+        $this->expectNotToPerformAssertions();
+        $subject = new Preg();
+
+        $subject->throwExceptions(true);
+        $subject->throwExceptions(false);
+        @$methodCaller($subject);
+    }
+
+    /**
+     * @return array<non-empty-string, array{
+     *             pattern: non-empty-string|array<non-empty-string>,
+     *             replacement: string|array<string>,
+     *             subject: string,
+     *             limit: int,
+     *             expect: string,
+     *         }>
+     */
+    public function providePregReplaceArgumentsAndExpectedResult(): array
+    {
+        return [
+            'string arguments' => [
+                'pattern' => '/a/',
+                'replacement' => 'fab',
+                'subject' => 'abba',
+                'limit' => -1,
+                'expect' => 'fabbbfab',
+            ],
+            'array pattern' => [
+                'pattern' => ['/a/', '/b/'],
+                'replacement' => 'fab',
+                'subject' => 'abba',
+                'limit' => -1,
+                'expect' => 'fafabfabfabfafab',
+            ],
+            'array pattern and replacement' => [
+                'pattern' => ['/a/', '/b/'],
+                'replacement' => ['fab', 'z'],
+                'subject' => 'abba',
+                'limit' => -1,
+                'expect' => 'fazzzfaz',
+            ],
+            'with limit' => [
+                'pattern' => '/a/',
+                'replacement' => 'fab',
+                'subject' => 'abba',
+                'limit' => 1,
+                'expect' => 'fabbba',
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param non-empty-string|non-empty-array<non-empty-string> $pattern
+     * @param string|non-empty-array<string> $replacement
+     *
+     * @dataProvider providePregReplaceArgumentsAndExpectedResult
+     */
+    public function replaceReplaces($pattern, $replacement, string $subject, int $limit, string $expectedResult): void
+    {
+        $testSubject = new Preg();
+
+        $result = $testSubject->replace($pattern, $replacement, $subject, $limit);
+
+        self::assertSame($expectedResult, $result);
+    }
+
+    /**
+     * @test
+     *
+     * @param non-empty-string|non-empty-array<non-empty-string> $pattern
+     * @param string|non-empty-array<string> $replacement
+     *
+     * @dataProvider providePregReplaceArgumentsAndExpectedResult
+     */
+    public function replaceCallbackReplaces(
+        $pattern,
+        $replacement,
+        string $subject,
+        int $limit,
+        string $expectedResult
+    ): void {
+        $this->replaceCallbackReplacement = $replacement;
+        $this->lastReplaceCallbackMatch = null;
+        $this->replaceCallbackReplacementIndex = -1;
+        $testSubject = new Preg();
+
+        $result = $testSubject->replaceCallback(
+            $pattern,
+            \Closure::fromCallable([$this, 'callbackForReplaceCallback']),
```

Comment:
Isn't the array also a callable, e.g. like this, so that we don't need the `Closure::fromCallable`?

`[$this, 'callbackForReplaceCallback']`

Reply:
It's not accessible to the method `replaceCallback` as it is not `public`, resulting in "[Argument passed to function must be callable, array given](https://stackoverflow.com/questions/43170785/argument-passed-to-function-must-be-callable-array-given)".  (The built in function `preg_replace_callback` circumvents the restriction in some internal way.)